src/Symfony/Component/Lock/Store/ZookeeperStore.php

```diff
@@ -0,0 +1,185 @@
+     *
+     * @throws LockConflictedException
+     * @throws LockAcquiringException
+     */
+    public function save(Key $key)
+    {
+        $resource = $this->getKeyResource($key);
+
+        if ($this->exists($key)) {
+            return;
+        }
+
+        $token = $this->getToken($key);
+
+        try {
+            $this->createNewLock($resource, $token);
+        } catch (LockConflictedException $ex) {
+            throw $ex;
+        } catch (\Throwable $throwable) {
+            throw new LockAcquiringException($throwable);
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @throws LockReleasingException
+     */
+    public function delete(Key $key)
+    {
+        if (!$this->exists($key)) {
+            return;
+        }
+        $resource = $this->getKeyResource($key);
+        try {
+            $this->zookeeper->delete($resource);
+        } catch (\Throwable $throwable) {
+            // For Zookeeper Ephemeral Nodes, the node will be deleted upon session death. But, if we want to unlock
+            // the lock before proceeding further in the session, the client should be aware of this
+            throw new LockReleasingException($throwable);
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function exists(Key $key): bool
+    {
+        $resource = $this->getKeyResource($key);
+        try {
+            return $this->zookeeper->get($resource) === $this->getToken($key);
+        } catch (\ZookeeperException $ex) {
+            return false;
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @throws NotSupportedException
+     */
+    public function waitAndSave(Key $key)
+    {
+        throw new NotSupportedException();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function putOffExpiration(Key $key, $ttl)
+    {
+        // This store doesn't support TTL. The lock nodes die at the end of a client's session if not released correctly
+    }
+
+    /**
+     * Create a zookeeper node.
+     *
+     * @param string $node  The node which needs to be created
+     * @param string $value The value to be assigned to a zookeeper node
+     *
+     * @throws InvalidArgumentException
+     * @throws LockConflictedException
+     * @throws LockAcquiringException
+     */
+    private function createNewLock(
+        string $node,
+        string $value = ''
+    ) {
+        // Default Node Permissions
+        $acl = array(array('perms' => \Zookeeper::PERM_ALL, 'scheme' => 'world', 'id' => 'anyone'));
+        // This ensures that the nodes are deleted when the client session to zookeeper server ends.
+        $type = \Zookeeper::EPHEMERAL;
+
+        // Client should just pass in a single node to be used as a lock.
+        $parts = array_filter(explode('/', $node));
+
+        if (\count($parts) > 1) {
+            throw new InvalidArgumentException('Lock nodes should be created at root level. Given: '.$node);
```

Comment:
I am not sure I entirely follow your concern here. If my response above doesn't cover this, could you please elaborate? Sorry and thanks!

Reply:
My concern is: This would be the first implementation of StoreInterface that throws an exception when the key does not match a pattern.

I suggest to replace `/` instead of throwing an exception.

```
$node = '/' . \strtr($key, ['/' => '~1', '~' => '~0']); # Like in JSON Pointer  https://tools.ietf.org/html/rfc6901#section-3

# or

$node = '/' . \strtr($key, ['/' => '-']).'-'.sha1($key); # sha1 is used to avoid conflict "foo-bar" == "foo/bar"
```

Reply:
Okay - yes this definitely makes sense to me and I think adding the sha will really be helpful. I have moved this logic into getKeyResource which is currently being called by all functions anyway. So right now if client passed a key which starts `/` even that will get converted to `-` which is acceptable I guess.